hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/impl/WeakReferenceThreadMap.java

```diff
@@ -48,7 +49,15 @@ public long currentThreadId() {
   }
 
   public V setForCurrentThread(V newVal) {
-    return put(currentThreadId(), newVal);
+    long id = currentThreadId();
+
+    // if the same object is already in the map, just return it.
+    WeakReference<V> ref = lookup(id);
+    if (ref != null && ref.get() == newVal) {
```

Comment:
if there's been a GC ref.get could be null. now, as we are doing == and not Object.equals that is safe.

we should mention that as a comment